File: MasteringSwift/MasteringSwift/chapters/chapter15/intermediate/IntermediateFunctional.swift
//
//  IntermediateFunctional.swift
//  MasteringSwift
//
//  Capítulo 15 — Programación funcional.
//  Nivel INTERMEDIO: map, filter, reduce sobre arrays.
//

import Foundation

func runIntermediateFunctionalExamples() {
    let numbers = [1, 2, 3, 4, 5, 6]

    let doubled = numbers.map { $0 * 2 }
    print("map (double): \(doubled)")

    let evens = numbers.filter { $0 % 2 == 0 }
    print("filter (even): \(evens)")

    let sum = numbers.reduce(0) { $0 + $1 }
    print("reduce (sum): \(sum)")

    let result = numbers.map { $0 * 2 }.filter { $0 > 5 }.reduce(0, +)
    print("chain (double, filter >5, sum): \(result)")
}

File: MasteringDart/lib/chapters/chapter15/intermediate/intermediate_functional.dart
// ignore_for_file: avoid_print
//
//  intermediate_functional.dart
//  MasteringDart
//
//  Capítulo 15 — Programación funcional.
//  Nivel INTERMEDIO: map, filter, reduce sobre listas.
//
//  Ejecutar con: dart run lib/chapters/chapter15/intermediate/intermediate_functional.dart

void runIntermediateFunctionalExamples() {
  const numbers = [1, 2, 3, 4, 5, 6];

  final doubled = numbers.map((n) => n * 2).toList();
  print('map (double): $doubled');

  final evens = numbers.where((n) => n % 2 == 0).toList();
  print('filter (even): $evens');

  final sum = numbers.reduce((a, b) => a + b);
  print('reduce (sum): $sum');

  final result = numbers
      .map((n) => n * 2)
      .where((n) => n > 5)
      .reduce((a, b) => a + b);
  print('chain (double, filter >5, sum): $result');
}

void main() {
  runIntermediateFunctionalExamples();
}
